src/dataStructures/InvertedIndex.java

```diff
@@ -0,0 +1,74 @@
+package dataStructures;
+
+import filter.WordValidator;
+import filter.normalizer.Normalizer;
+import filter.stemmer.Stemmer;
+import filter.tokenizer.Tokenizer;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class InvertedIndex {
+    private final Map<String, Set<String>> engine;
+    private final Normalizer normalizer;
+    private final Tokenizer tokenizer;
+    private WordValidator wordValidator;
+    private boolean doStem;
+
+    public InvertedIndex(Tokenizer tokenizer, Normalizer normalizer) {
+        this.engine = new HashMap<>();
+        this.tokenizer = tokenizer;
+        this.normalizer = normalizer;
+    }
+
+    public void setDoStem(boolean doStem) {
+        this.doStem = doStem;
+    }
+
+    public void setWordValidator(WordValidator validator) {
+        this.wordValidator = validator;
+    }
+
+    public WordValidator getWordValidator() {
+        return wordValidator;
+    }
+
+    public Normalizer getNormalizer() {
+        return normalizer;
+    }
+
+    private void addWordToEngine(String root, String fileName) {
+        if (!engine.containsKey(root)) {
+            engine.put(root, new HashSet<>());
+        }
+        engine.get(root).add(fileName);
+    }
+
+    public String getWordRoot(String word) {
```

Comment:
each time you use the `getWordRoot` method, you are creating a new instance of Stemmer()